### DISCUSSION TITLE
How about const class

### DISCUSSION THREAD

**Discussion by lindexi** [dotnet/csharplang#4755] [Language Ideas]
How about the const class:

```csharp
public const class Foo
{
      string F1 = "F1";
      string F2 = "F2";
}
```

Equal to:

```csharp
    public static class Foo
    {
        public const string F1 = "F1";
        public const string F2 = "F2";
    }
```

**Comment by huoyaoyuan** [comment]
The syntax is counter-institutive. In C#, `const` is about immutability. Seeing the title `const class` implies some sort of immutable class for me.

**Reply by cheong00** [reply]
I think he want to talk about the idea of something like "string based enum".

The patent hold by Oracle is a major concern.

#274 

Also see the idea of "enum class" in #144

**Reply by lindexi** [reply]
@cheong00 I do not think so. 

```
const class Foo
{
           string F1 = "F1";
           int      F3 = 3;
}
```

**Reply by bernd5** [reply]
Just like "static class" forces that every member is static - in such a "const class" every member would be const...

Right?

**Reply by lindexi** [reply]
@bernd5 Yes

**Reply by Joe4evr** [reply]
Even if this was done, I sincerely doubt the implementation would let you omit the `const` modifier on the member declarations for the same reason that you don't get to omit the `static` modifier on members inside of static classes, so this would end up saving you a grand total of 1 character.

In that light, your options are:
1. Just say `static class Constants` and suck it up
2. If you *really* want to force a type to only contain `const` members, write a custom analyzer to enforce that semantic; that's what analyzers are meant for.
